test/reduce.cpp

```diff
@@ -312,6 +318,969 @@ CPP_REDUCE_TESTS(anyTrue, any_true, float, unsigned char);
 CPP_REDUCE_TESTS(allTrue, all_true, float, unsigned char);
 CPP_REDUCE_TESTS(count, count, float, unsigned);
 
+//
+// Reduce By Key tests
+//
+template <typename Tk, typename Tv>
+struct reduce_by_key_params {
+    string testname_;
+    vector<Tk> iKeys_;
+    vector<Tv> iVals_;
+    vector<Tk> oKeys_;
+    vector<Tv> oVals_;
+
+    reduce_by_key_params(vector<Tk> ikeys, vector<Tv> ivals, vector<Tk> okeys,
+                         vector<Tv> ovals, string testname)
+        : iKeys_(ikeys)
+        , iVals_(ivals)
+        , oKeys_(okeys)
+        , oVals_(ovals)
+        , testname_(testname) {}
+};
+
+class ReduceByKey_if
+    : public ::testing::TestWithParam<reduce_by_key_params<int, float> > {};
+class ReduceByKey_uf
+    : public ::testing::TestWithParam<reduce_by_key_params<unsigned, float> > {
+};
+class ReduceByKey_id
+    : public ::testing::TestWithParam<reduce_by_key_params<int, double> > {};
+class ReduceByKey_iacf
+    : public ::testing::TestWithParam<reduce_by_key_params<int, cfloat> > {};
+class ReduceByKey_iacd
+    : public ::testing::TestWithParam<reduce_by_key_params<int, cdouble> > {};
+
+class ReduceByKey_dim_if
+    : public ::testing::TestWithParam<reduce_by_key_params<int, float> > {};
+class ReduceByKey_dim_uf
+    : public ::testing::TestWithParam<reduce_by_key_params<unsigned, float> > {
+};
+class ReduceByKey_dim_id
+    : public ::testing::TestWithParam<reduce_by_key_params<int, double> > {};
+class ReduceByKey_dim_iacf
+    : public ::testing::TestWithParam<reduce_by_key_params<int, cfloat> > {};
+class ReduceByKey_dim_iacd
+    : public ::testing::TestWithParam<reduce_by_key_params<int, cdouble> > {};
+
```

Comment:
Perhaps, the names could use `af_dtype` names instead of this new convention. I didn't get it untill I realized the pattern across all class definitions.

Or some uniform naming convention like blas fns have 
- `s` float
- `d` double
- `c` complex float
- `z` complex double
And then we add new 
- `i` integer
- `u` unsigned integer
That way, all names would be of similar pattern `ReduceByKey_[if | uf | sf | df | cf | zf  ]` etc.